File: src/decision.c
#include "decision.h"

DataStructure decideStructure(const Answers *a) {
    if (a->needsLIFO) return DS_STACK;
    if (a->needsFIFO) return DS_QUEUE;
    if (a->needsPriority) return DS_HEAP;
    if (a->hasGraphRelationships) return DS_GRAPH;
    if (a->needsOrdering) return DS_BST;
    if (a->wantsRandomAccess) return DS_ARRAY;
    if (a->manyMiddleInsertions) return DS_LINKED_LIST;
    return DS_ARRAY;
}

const char *dsName(DataStructure ds) {
    switch (ds) {
        case DS_ARRAY: return "Array (arreglo)";
        case DS_LINKED_LIST: return "Linked List (lista ligada)";
        case DS_STACK: return "Stack (pila)";
        case DS_QUEUE: return "Queue (cola)";
        case DS_BST: return "Binary Search Tree (BST)";
        case DS_HEAP: return "Heap (priority queue)";
        case DS_GRAPH: return "Graph (grafo)";
        default: return "Desconocida";
    }
}

File: src/decision.py
from enum import Enum, auto
from .questions import Answers


class DataStructure(Enum):
    ARRAY = auto()
    LINKED_LIST = auto()
    STACK = auto()
    QUEUE = auto()
    BST = auto()
    HEAP = auto()
    GRAPH = auto()


def ds_name(ds: DataStructure) -> str:
    return {
        DataStructure.ARRAY: "Array",
        DataStructure.LINKED_LIST: "Linked List",
        DataStructure.STACK: "Stack",
        DataStructure.QUEUE: "Queue",
        DataStructure.BST: "Binary Search Tree (BST)",
        DataStructure.HEAP: "Heap / Priority Queue",
        DataStructure.GRAPH: "Graph",
    }[ds]


def decide_structure(ans: Answers) -> DataStructure:
    if ans.wants_random_access:
        return DataStructure.ARRAY

    if ans.needs_lifo:
        return DataStructure.STACK

    if ans.needs_fifo:
        return DataStructure.QUEUE

    if ans.many_middle_insertions:
        return DataStructure.LINKED_LIST

    if ans.needs_ordering:
        return DataStructure.BST

    if ans.needs_priority:
        return DataStructure.HEAP

    if ans.has_graph_relationships:
        return DataStructure.GRAPH

    return DataStructure.ARRAY
